File: src/memory/conflict_resolver.py
"""
Conflict Resolver Module

Extracted from unified_memory.py with enhancements:
- ConflictResolutionStrategy enum for strategy selection
- Enhanced semantic similarity detection
- Conflict resolution with configurable strategies

Strategies:
- AUTO: Newer information supersedes older (default)
- KEEP_OLD: Preserve existing information
- KEEP_NEW: Replace with new information
- MERGE: Combine both versions
- MANUAL: Defer to user decision
"""

import logging
from dataclasses import dataclass, field
from datetime import datetime
from enum import Enum
from typing import Any, Dict, List, Optional, Protocol, Tuple, runtime_checkable

logger = logging.getLogger("niko-conflict-resolver")


class ConflictResolutionStrategy(Enum):
    """Strategies for resolving memory conflicts."""
    AUTO = "auto"           # Automatic: newer supersedes older
    KEEP_OLD = "keep_old"   # Preserve existing information
    KEEP_NEW = "keep_new"   # Replace with new information
    MERGE = "merge"         # Combine both versions
    MANUAL = "manual"       # Defer to user decision


class ConflictType(Enum):
    """Types of detected conflicts."""
    CONTRADICTION = "contradiction"   # Direct semantic contradiction
    TEMPORAL = "temporal"             # Time-based conflict
    DUPLICATE = "duplicate"           # Near-duplicate content
    UPDATE = "update"                 # Information update
    AMBIGUOUS = "ambiguous"           # Unclear conflict


@dataclass
class ConflictInfo:
    """Information about a detected conflict."""
    id: str
    content: str
    valid_from: Optional[str] = None
    valid_until: Optional[str] = None
    importance: float = 0.5
    conflict_type: ConflictType = ConflictType.CONTRADICTION
    similarity_score: float = 0.0
    metadata: Dict[str, Any] = field(default_factory=dict)


@dataclass
class ResolutionResult:
    """Result of conflict resolution."""
    action: str  # "update", "reject", "merge", "defer"
    kept_ids: List[str] = field(default_factory=list)
    obsolete_ids: List[str] = field(default_factory=list)
    merged_content: Optional[str] = None
    reason: str = ""
    requires_manual: bool = False
    metadata: Dict[str, Any] = field(default_factory=dict)


@runtime_checkable
class IConflictResolver(Protocol):
    """Protocol for conflict resolution implementations."""

    async def check(
        self,
        content: str,
        entity_id: Optional[str] = None
    ) -> List[ConflictInfo]:
        """
        Check for potential conflicts.

        Args:
            content: New content to check.
            entity_id: Entity ID to scope the check.

        Returns:
            List of detected conflicts.
        """
        ...

    async def resolve(
        self,
        content: str,
        conflicts: List[ConflictInfo],
        strategy: ConflictResolutionStrategy = ConflictResolutionStrategy.AUTO
    ) -> ResolutionResult:
        """
        Resolve detected conflicts.

        Args:
            content: New content being added.
            conflicts: List of detected conflicts.
            strategy: Resolution strategy to use.

        Returns:
            ResolutionResult with action to take.
        """
        ...


class ConflictResolver:
    """
    Conflict detection and resolution engine.

    Detects contradictions, duplicates, and temporal conflicts
    in memory content and provides resolution strategies.
    """

    # Negation pairs for contradiction detection
    NEGATION_PAIRS = [
        ("is", "is not"), ("are", "are not"),
        ("was", "was not"), ("were", "were not"),
        ("has", "has not"), ("have", "have not"),
        ("can", "cannot"), ("will", "will not"),
        ("does", "does not"), ("do", "do not"),
        ("alive", "dead"), ("true", "false"),
        ("yes", "no"), ("accept", "reject"),
        ("love", "hate"), ("friend", "enemy"),
        ("success", "failure"), ("win", "lose"),
    ]

    def __init__(self, db_connection=None, similarity_threshold: float = 0.85):
        """
        Initialize ConflictResolver.

        Args:
            db_connection: Optional database connection for querying existing memories.
            similarity_threshold: Threshold for considering content as duplicate.
        """
        self.db = db_connection
        self.similarity_threshold = similarity_threshold
        self._embedder = None
        logger.info("ConflictResolver initialized")

    def set_db_connection(self, db_connection) -> None:
        """Set or update database connection."""
        self.db = db_connection

    def set_embedder(self, embedder) -> None:
        """Set embedding engine for semantic similarity."""
        self._embedder = embedder

    async def check(
        self,
        content: str,
        entity_id: Optional[str] = None
    ) -> List[ConflictInfo]:
        """
        Check for potential conflicts with existing memories.

        Args:
            content: New content to check.
            entity_id: Entity ID to scope the check.

        Returns:
            List of detected conflicts.
        """
        if not entity_id or not self.db:
            return []

        try:
            cursor = self.db.execute("""
                SELECT id, content, valid_from, valid_until, importance
                FROM memories
                WHERE entity_id = ?
                AND superseded_by IS NULL
                AND (valid_until IS NULL OR valid_until > datetime('now'))
            """, (entity_id,))

            conflicts = []
            rows = cursor.fetchall()
            for row in rows:
                existing_content = row[1]
                conflict_type, similarity = self._analyze_conflict(content, existing_content)

                if conflict_type is not None:
                    conflicts.append(ConflictInfo(
                        id=row[0],
                        content=existing_content,
                        valid_from=row[2],
                        valid_until=row[3],
                        importance=row[4] or 0.5,
                        conflict_type=conflict_type,
                        similarity_score=similarity
                    ))

            return conflicts

        except Exception as e:
            logger.error(f"Conflict check failed: {e}")
            return []

    def _analyze_conflict(
        self,
        new_content: str,
        existing_content: str
    ) -> Tuple[Optional[ConflictType], float]:
        """
        Analyze potential conflict between contents.

        Returns:
            Tuple of (conflict_type, similarity_score) or (None, 0.0) if no conflict.
        """
        # Check for contradiction
        if self._is_contradictory(new_content, existing_content):
            return ConflictType.CONTRADICTION, 0.0

        # Check for semantic similarity (duplicate/update)
        similarity = self._calculate_similarity(new_content, existing_content)

        if similarity >= self.similarity_threshold:
            return ConflictType.DUPLICATE, similarity
        elif similarity >= 0.6:
            return ConflictType.UPDATE, similarity

        return None, 0.0

    def _is_contradictory(self, content_a: str, content_b: str) -> bool:
        """
        Detect if two contents are contradictory.

        Uses negation pair detection and semantic analysis.
        """
        a_lower = content_a.lower()
        b_lower = content_b.lower()

        # Check negation pairs
        for pos, neg in self.NEGATION_PAIRS:
            if (pos in a_lower and neg in b_lower) or \
               (neg in a_lower and pos in b_lower):
                # Verify they're talking about the same subject
                if self._share_subject(content_a, content_b):
                    return True

        return False

    def _share_subject(self, content_a: str, content_b: str) -> bool:
        """Check if two contents share a common subject."""
        # Extract nouns/proper nouns (simplified)
        import re

        def extract_words(text: str) -> set:
            words = set(re.findall(r'\b[A-Za-z]+\b', text.lower()))
            # Filter out common words
            stopwords = {'the', 'a', 'an', 'is', 'are', 'was', 'were', 'be',
                        'been', 'being', 'have', 'has', 'had', 'do', 'does',
                        'did', 'will', 'would', 'could', 'should', 'may',
                        'might', 'must', 'shall', 'can', 'need', 'to', 'of',
                        'in', 'for', 'on', 'with', 'at', 'by', 'from', 'as',
                        'into', 'through', 'during', 'before', 'after',
                        'above', 'below', 'between', 'under', 'again',
                        'further', 'then', 'once', 'and', 'but', 'or', 'nor',
                        'so', 'yet', 'both', 'either', 'neither', 'not',
                        'only', 'own', 'same', 'than', 'too', 'very', 'just'}
            return words - stopwords

        words_a = extract_words(content_a)
        words_b = extract_words(content_b)

        # Check for significant overlap
        overlap = words_a & words_b
        min_size = min(len(words_a), len(words_b))

        if min_size == 0:
            return False

        return len(overlap) / min_size >= 0.3

    def _calculate_similarity(self, content_a: str, content_b: str) -> float:
        """
        Calculate semantic similarity between contents.

        Uses embedding similarity if available, falls back to lexical.
        """
        if self._embedder is not None:
            try:
                vec_a = self._embedder.embed(content_a)
                vec_b = self._embedder.embed(content_b)
                return self._embedder.similarity(vec_a, vec_b)
            except Exception as e:
                logger.debug(f"Embedding similarity failed: {e}")

        # Fallback: Jaccard similarity
        words_a = set(content_a.lower().split())
        words_b = set(content_b.lower().split())

        if not words_a or not words_b:
            return 0.0

        intersection = len(words_a & words_b)
        union = len(words_a | words_b)

        return intersection / union if union > 0 else 0.0

    async def resolve(
        self,
        content: str,
        conflicts: List[ConflictInfo],
        strategy: ConflictResolutionStrategy = ConflictResolutionStrategy.AUTO
    ) -> ResolutionResult:
        """
        Resolve detected conflicts using specified strategy.

        Args:
            content: New content being added.
            conflicts: List of detected conflicts.
            strategy: Resolution strategy to use.

        Returns:
            ResolutionResult with action to take.
        """
        if not conflicts:
            return ResolutionResult(
                action="accept",
                reason="No conflicts detected"
            )

        if strategy == ConflictResolutionStrategy.AUTO:
            return await self._resolve_auto(content, conflicts)
        elif strategy == ConflictResolutionStrategy.KEEP_OLD:
            return self._resolve_keep_old(conflicts)
        elif strategy == ConflictResolutionStrategy.KEEP_NEW:
            return self._resolve_keep_new(conflicts)
        elif strategy == ConflictResolutionStrategy.MERGE:
            return self._resolve_merge(content, conflicts)
        elif strategy == ConflictResolutionStrategy.MANUAL:
            return self._resolve_manual(content, conflicts)
        else:
            return await self._resolve_auto(content, conflicts)

    async def _resolve_auto(
        self,
        content: str,
        conflicts: List[ConflictInfo]
    ) -> ResolutionResult:
        """
        Automatic resolution: newer information supersedes older.

        For duplicates, keeps the most important.
        For contradictions, uses recency.
        """
        obsolete_ids = []

        for conflict in conflicts:
            if conflict.conflict_type == ConflictType.DUPLICATE:
                # Keep higher importance
                if conflict.importance < 0.5:
                    obsolete_ids.append(conflict.id)
            elif conflict.conflict_type in (ConflictType.CONTRADICTION, ConflictType.UPDATE):
                # Newer supersedes older
                obsolete_ids.append(conflict.id)

        return ResolutionResult(
            action="update",
            obsolete_ids=obsolete_ids,
            reason="Newer information supersedes older (auto strategy)"
        )

    def _resolve_keep_old(self, conflicts: List[ConflictInfo]) -> ResolutionResult:
        """Keep existing information, reject new content."""
        return ResolutionResult(
            action="reject",
            kept_ids=[c.id for c in conflicts],
            reason="Keeping existing information (keep_old strategy)"
        )

    def _resolve_keep_new(self, conflicts: List[ConflictInfo]) -> ResolutionResult:
        """Replace with new information."""
        return ResolutionResult(
            action="update",
            obsolete_ids=[c.id for c in conflicts],
            reason="Replacing with new information (keep_new strategy)"
        )

    def _resolve_merge(
        self,
        content: str,
        conflicts: List[ConflictInfo]
    ) -> ResolutionResult:
        """Merge old and new content."""
        # Find the most relevant conflict to merge with
        primary_conflict = max(conflicts, key=lambda c: c.importance)

        # Create merged content
        merged = f"{primary_conflict.content}\n\n[Updated]: {content}"

        return ResolutionResult(
            action="merge",
            obsolete_ids=[c.id for c in conflicts],
            merged_content=merged,
            reason="Merged old and new information (merge strategy)"
        )

    def _resolve_manual(
        self,
        content: str,
        conflicts: List[ConflictInfo]
    ) -> ResolutionResult:
        """Defer to manual resolution."""
        return ResolutionResult(
            action="defer",
            requires_manual=True,
            reason="Manual resolution required",
            metadata={
                "new_content": content,
                "conflicts": [
                    {
                        "id": c.id,
                        "content": c.content,
                        "type": c.conflict_type.value,
                        "similarity": c.similarity_score
                    }
                    for c in conflicts
                ]
            }
        )

    def detect_all_conflicts(
        self,
        memories: List[Dict[str, Any]]
    ) -> List[Tuple[str, str, ConflictType]]:
        """
        Detect all conflicts within a set of memories.

        Args:
            memories: List of memory dicts with 'id' and 'content'.

        Returns:
            List of (id_a, id_b, conflict_type) tuples.
        """
        conflicts = []

        for i, mem_a in enumerate(memories):
            for mem_b in memories[i+1:]:
                conflict_type, _ = self._analyze_conflict(
                    mem_a.get("content", ""),
                    mem_b.get("content", "")
                )
                if conflict_type is not None:
                    conflicts.append((
                        mem_a.get("id", ""),
                        mem_b.get("id", ""),
                        conflict_type
                    ))

        return conflicts


# Singleton instance
_conflict_resolver: Optional[ConflictResolver] = None


def get_conflict_resolver(
    db_connection=None,
    similarity_threshold: float = 0.85
) -> ConflictResolver:
    """Get or create ConflictResolver singleton."""
    global _conflict_resolver
    if _conflict_resolver is None:
        _conflict_resolver = ConflictResolver(db_connection, similarity_threshold)
    elif db_connection is not None and _conflict_resolver.db is None:
        _conflict_resolver.set_db_connection(db_connection)
    return _conflict_resolver


def reset_conflict_resolver() -> None:
    """Reset ConflictResolver singleton (for testing)."""
    global _conflict_resolver
    _conflict_resolver = None

File: src-ts/memory/conflict-resolver.ts
/**
 * Conflict Resolver Module
 *
 * Extracted from unified_memory.py with enhancements:
 * - ConflictResolutionStrategy enum for strategy selection
 * - Enhanced semantic similarity detection
 * - Conflict resolution with configurable strategies
 *
 * Strategies:
 * - AUTO: Newer information supersedes older (default)
 * - KEEP_OLD: Preserve existing information
 * - KEEP_NEW: Replace with new information
 * - MERGE: Combine both versions
 * - MANUAL: Defer to user decision
 */

import { createLogger } from "../logger/index.js";

const _log = createLogger("conflict");

/** Strategies for resolving memory conflicts. */
export enum ConflictResolutionStrategy {
  AUTO = "auto", // Automatic: newer supersedes older
  KEEP_OLD = "keep_old", // Preserve existing information
  KEEP_NEW = "keep_new", // Replace with new information
  MERGE = "merge", // Combine both versions
  MANUAL = "manual", // Defer to user decision
}

/** Types of detected conflicts. */
export enum ConflictType {
  CONTRADICTION = "contradiction", // Direct semantic contradiction
  TEMPORAL = "temporal", // Time-based conflict
  DUPLICATE = "duplicate", // Near-duplicate content
  UPDATE = "update", // Information update
  AMBIGUOUS = "ambiguous", // Unclear conflict
}

/** Information about a detected conflict. */
export class ConflictInfo {
  id: string;
  content: string;
  validFrom: string | null;
  validUntil: string | null;
  importance: number;
  conflictType: ConflictType;
  similarityScore: number;
  metadata: Record<string, unknown>;

  constructor(params: {
    id: string;
    content: string;
    validFrom?: string | null;
    validUntil?: string | null;
    importance?: number;
    conflictType?: ConflictType;
    similarityScore?: number;
    metadata?: Record<string, unknown>;
  }) {
    this.id = params.id;
    this.content = params.content;
    this.validFrom = params.validFrom ?? null;
    this.validUntil = params.validUntil ?? null;
    this.importance = params.importance ?? 0.5;
    this.conflictType = params.conflictType ?? ConflictType.CONTRADICTION;
    this.similarityScore = params.similarityScore ?? 0.0;
    this.metadata = params.metadata ?? {};
  }
}

/** Result of conflict resolution. */
export class ResolutionResult {
  action: string; // "update", "reject", "merge", "defer"
  keptIds: string[];
  obsoleteIds: string[];
  mergedContent: string | null;
  reason: string;
  requiresManual: boolean;
  metadata: Record<string, unknown>;

  constructor(params: {
    action: string;
    keptIds?: string[];
    obsoleteIds?: string[];
    mergedContent?: string | null;
    reason?: string;
    requiresManual?: boolean;
    metadata?: Record<string, unknown>;
  }) {
    this.action = params.action;
    this.keptIds = params.keptIds ?? [];
    this.obsoleteIds = params.obsoleteIds ?? [];
    this.mergedContent = params.mergedContent ?? null;
    this.reason = params.reason ?? "";
    this.requiresManual = params.requiresManual ?? false;
    this.metadata = params.metadata ?? {};
  }
}

/** Interface for conflict resolution implementations. */
export interface IConflictResolver {
  /**
   * Check for potential conflicts.
   *
   * @param content - New content to check.
   * @param entityId - Entity ID to scope the check.
   * @returns List of detected conflicts.
   */
  check(
    content: string,
    entityId?: string | null,
    scope?: {
      userId?: string | null;
      projectId?: string | null;
      sessionId?: string | null;
    }
  ): Promise<ConflictInfo[]>;

  /**
   * Resolve detected conflicts.
   *
   * @param content - New content being added.
   * @param conflicts - List of detected conflicts.
   * @param strategy - Resolution strategy to use.
   * @returns ResolutionResult with action to take.
   */
  resolve(
    content: string,
    conflicts: ConflictInfo[],
    strategy?: ConflictResolutionStrategy
  ): Promise<ResolutionResult>;
}

/** Interface for database connections used by ConflictResolver */
export interface ConflictDbConnection {
  execute(sql: string, params: unknown[]): { fetchAll(): unknown[][] };
}

/** Interface for embedder used by ConflictResolver */
export interface ConflictEmbedder {
  embed(text: string): number[];
  similarity(vecA: number[], vecB: number[]): number;
}

/**
 * Conflict detection and resolution engine.
 *
 * Detects contradictions, duplicates, and temporal conflicts
 * in memory content and provides resolution strategies.
 */
export class ConflictResolver implements IConflictResolver {
  /** Negation pairs for contradiction detection */
  static readonly NEGATION_PAIRS: [string, string][] = [
    ["is", "is not"],
    ["are", "are not"],
    ["was", "was not"],
    ["were", "were not"],
    ["has", "has not"],
    ["have", "have not"],
    ["can", "cannot"],
    ["will", "will not"],
    ["does", "does not"],
    ["do", "do not"],
    ["alive", "dead"],
    ["true", "false"],
    ["yes", "no"],
    ["accept", "reject"],
    ["love", "hate"],
    ["friend", "enemy"],
    ["success", "failure"],
    ["win", "lose"],
  ];

  db: ConflictDbConnection | null;
  similarityThreshold: number;
  private _embedder: ConflictEmbedder | null;

  constructor(
    dbConnection?: ConflictDbConnection | null,
    similarityThreshold: number = 0.85
  ) {
    this.db = dbConnection ?? null;
    this.similarityThreshold = similarityThreshold;
    this._embedder = null;
    _log.info("ConflictResolver initialized");
  }

  /** Set or update database connection. */
  setDbConnection(dbConnection: ConflictDbConnection): void {
    this.db = dbConnection;
  }

  /** Set embedding engine for semantic similarity. */
  setEmbedder(embedder: ConflictEmbedder): void {
    this._embedder = embedder;
  }

  /**
   * Check for potential conflicts with existing memories.
   *
   * @param content - New content to check.
   * @param entityId - Entity ID to scope the check.
   * @returns List of detected conflicts.
   */
  async check(
    content: string,
    entityId?: string | null,
    scope: {
      userId?: string | null;
      projectId?: string | null;
      sessionId?: string | null;
    } = {}
  ): Promise<ConflictInfo[]> {
    if (!entityId || !this.db) {
      return [];
    }

    try {
      let sql = `
        SELECT id, content, valid_from, valid_until, importance
        FROM memories
        WHERE entity_id = ?
        AND superseded_by IS NULL
        AND (valid_until IS NULL OR valid_until > datetime('now'))
      `;
      const sqlParams: unknown[] = [entityId];

      if (scope.userId !== undefined && scope.userId !== null) {
        sql += ' AND user_id = ?';
        sqlParams.push(scope.userId);
      }
      if (scope.projectId !== undefined && scope.projectId !== null) {
        sql += ' AND project_id = ?';
        sqlParams.push(scope.projectId);
      }
      if (scope.sessionId !== undefined && scope.sessionId !== null) {
        sql += ' AND session_id = ?';
        sqlParams.push(scope.sessionId);
      }

      const cursor = this.db.execute(
        sql,
        sqlParams
      );

      const conflicts: ConflictInfo[] = [];
      const rows = cursor.fetchAll();
      for (const row of rows) {
        const existingContent = row[1] as string;
        const [conflictType, similarity] = this._analyzeConflict(
          content,
          existingContent
        );

        if (conflictType !== null) {
          conflicts.push(
            new ConflictInfo({
              id: row[0] as string,
              content: existingContent,
              validFrom: (row[2] as string) ?? null,
              validUntil: (row[3] as string) ?? null,
              importance: (row[4] as number) ?? 0.5,
              conflictType,
              similarityScore: similarity,
            })
          );
        }
      }

      return conflicts;
    } catch (e) {
      _log.error(`Conflict check failed: ${e}`);
      return [];
    }
  }

  /**
   * Analyze potential conflict between contents.
   *
   * @returns Tuple of [conflictType, similarityScore] or [null, 0.0] if no conflict.
   */
  private _analyzeConflict(
    newContent: string,
    existingContent: string
  ): [ConflictType | null, number] {
    // Check for contradiction
    if (this._isContradictory(newContent, existingContent)) {
      return [ConflictType.CONTRADICTION, 0.0];
    }

    // Check for semantic similarity (duplicate/update)
    const similarity = this._calculateSimilarity(newContent, existingContent);

    if (similarity >= this.similarityThreshold) {
      return [ConflictType.DUPLICATE, similarity];
    } else if (similarity >= 0.6) {
      return [ConflictType.UPDATE, similarity];
    }

    return [null, 0.0];
  }

  /**
   * Detect if two contents are contradictory.
   *
   * Uses negation pair detection and semantic analysis.
   */
  private _isContradictory(contentA: string, contentB: string): boolean {
    const aLower = contentA.toLowerCase();
    const bLower = contentB.toLowerCase();

    // Check negation pairs
    for (const [pos, neg] of ConflictResolver.NEGATION_PAIRS) {
      if (
        (aLower.includes(pos) && bLower.includes(neg)) ||
        (aLower.includes(neg) && bLower.includes(pos))
      ) {
        // Verify they're talking about the same subject
        if (this._shareSubject(contentA, contentB)) {
          return true;
        }
      }
    }

    return false;
  }

  /** Check if two contents share a common subject. */
  private _shareSubject(contentA: string, contentB: string): boolean {
    const stopwords = new Set([
      "the", "a", "an", "is", "are", "was", "were", "be",
      "been", "being", "have", "has", "had", "do", "does",
      "did", "will", "would", "could", "should", "may",
      "might", "must", "shall", "can", "need", "to", "of",
      "in", "for", "on", "with", "at", "by", "from", "as",
      "into", "through", "during", "before", "after",
      "above", "below", "between", "under", "again",
      "further", "then", "once", "and", "but", "or", "nor",
      "so", "yet", "both", "either", "neither", "not",
      "only", "own", "same", "than", "too", "very", "just",
    ]);

    function extractWords(text: string): Set<string> {
      const matches = text.toLowerCase().match(/\b[a-z]+\b/g) ?? [];
      return new Set(matches.filter((w) => !stopwords.has(w)));
    }

    const wordsA = extractWords(contentA);
    const wordsB = extractWords(contentB);

    // Check for significant overlap
    const overlap = new Set([...wordsA].filter((w) => wordsB.has(w)));
    const minSize = Math.min(wordsA.size, wordsB.size);

    if (minSize === 0) {
      return false;
    }

    return overlap.size / minSize >= 0.3;
  }

  /**
   * Calculate semantic similarity between contents.
   *
   * Uses embedding similarity if available, falls back to lexical.
   */
  private _calculateSimilarity(contentA: string, contentB: string): number {
    if (this._embedder !== null) {
      try {
        const vecA = this._embedder.embed(contentA);
        const vecB = this._embedder.embed(contentB);
        return this._embedder.similarity(vecA, vecB);
      } catch (e) {
        _log.warn('Embedding similarity failed, falling back to Jaccard', { detail: e });
      }
    }

    // Fallback: Jaccard similarity
    const wordsA = new Set(contentA.toLowerCase().split(/\s+/).filter(Boolean));
    const wordsB = new Set(contentB.toLowerCase().split(/\s+/).filter(Boolean));

    if (wordsA.size === 0 || wordsB.size === 0) {
      return 0.0;
    }

    const intersection = new Set([...wordsA].filter((w) => wordsB.has(w)));
    const union = new Set([...wordsA, ...wordsB]);

    return intersection.size / union.size;
  }

  /**
   * Resolve detected conflicts using specified strategy.
   *
   * @param content - New content being added.
   * @param conflicts - List of detected conflicts.
   * @param strategy - Resolution strategy to use.
   * @returns ResolutionResult with action to take.
   */
  async resolve(
    content: string,
    conflicts: ConflictInfo[],
    strategy: ConflictResolutionStrategy = ConflictResolutionStrategy.AUTO
  ): Promise<ResolutionResult> {
    if (conflicts.length === 0) {
      return new ResolutionResult({
        action: "accept",
        reason: "No conflicts detected",
      });
    }

    switch (strategy) {
      case ConflictResolutionStrategy.AUTO:
        return this._resolveAuto(content, conflicts);
      case ConflictResolutionStrategy.KEEP_OLD:
        return this._resolveKeepOld(conflicts);
      case ConflictResolutionStrategy.KEEP_NEW:
        return this._resolveKeepNew(conflicts);
      case ConflictResolutionStrategy.MERGE:
        return this._resolveMerge(content, conflicts);
      case ConflictResolutionStrategy.MANUAL:
        return this._resolveManual(content, conflicts);
      default:
        return this._resolveAuto(content, conflicts);
    }
  }

  /**
   * Automatic resolution: newer information supersedes older.
   *
   * For duplicates, keeps the most important.
   * For contradictions, uses recency.
   */
  private async _resolveAuto(
    _content: string,
    conflicts: ConflictInfo[]
  ): Promise<ResolutionResult> {
    const obsoleteIds: string[] = [];

    for (const conflict of conflicts) {
      if (conflict.conflictType === ConflictType.DUPLICATE) {
        // Keep higher importance
        if (conflict.importance < 0.5) {
          obsoleteIds.push(conflict.id);
        }
      } else if (
        conflict.conflictType === ConflictType.CONTRADICTION ||
        conflict.conflictType === ConflictType.UPDATE
      ) {
        // Newer supersedes older
        obsoleteIds.push(conflict.id);
      }
    }

    return new ResolutionResult({
      action: "update",
      obsoleteIds,
      reason: "Newer information supersedes older (auto strategy)",
    });
  }

  /** Keep existing information, reject new content. */
  private _resolveKeepOld(conflicts: ConflictInfo[]): ResolutionResult {
    return new ResolutionResult({
      action: "reject",
      keptIds: conflicts.map((c) => c.id),
      reason: "Keeping existing information (keep_old strategy)",
    });
  }

  /** Replace with new information. */
  private _resolveKeepNew(conflicts: ConflictInfo[]): ResolutionResult {
    return new ResolutionResult({
      action: "update",
      obsoleteIds: conflicts.map((c) => c.id),
      reason: "Replacing with new information (keep_new strategy)",
    });
  }

  /** Merge old and new content. */
  private _resolveMerge(
    content: string,
    conflicts: ConflictInfo[]
  ): ResolutionResult {
    // Find the most relevant conflict to merge with
    const primaryConflict = conflicts.reduce((max, c) =>
      c.importance > max.importance ? c : max
    );

    // Create merged content
    const merged = `${primaryConflict.content}\n\n[Updated]: ${content}`;

    return new ResolutionResult({
      action: "merge",
      obsoleteIds: conflicts.map((c) => c.id),
      mergedContent: merged,
      reason: "Merged old and new information (merge strategy)",
    });
  }

  /** Defer to manual resolution. */
  private _resolveManual(
    content: string,
    conflicts: ConflictInfo[]
  ): ResolutionResult {
    return new ResolutionResult({
      action: "defer",
      requiresManual: true,
      reason: "Manual resolution required",
      metadata: {
        new_content: content,
        conflicts: conflicts.map((c) => ({
          id: c.id,
          content: c.content,
          type: c.conflictType,
          similarity: c.similarityScore,
        })),
      },
    });
  }

  /**
   * Detect all conflicts within a set of memories.
   *
   * @param memories - List of memory objects with 'id' and 'content'.
   * @returns List of [idA, idB, conflictType] tuples.
   */
  detectAllConflicts(
    memories: Array<{ id?: string; content?: string }>
  ): [string, string, ConflictType][] {
    const conflicts: [string, string, ConflictType][] = [];

    for (let i = 0; i < memories.length; i++) {
      for (let j = i + 1; j < memories.length; j++) {
        const [conflictType] = this._analyzeConflict(
          memories[i].content ?? "",
          memories[j].content ?? ""
        );
        if (conflictType !== null) {
          conflicts.push([
            memories[i].id ?? "",
            memories[j].id ?? "",
            conflictType,
          ]);
        }
      }
    }

    return conflicts;
  }
}

// Singleton instance
let _conflictResolver: ConflictResolver | null = null;

/** Get or create ConflictResolver singleton. */
export function getConflictResolver(
  dbConnection?: ConflictDbConnection | null,
  similarityThreshold?: number
): ConflictResolver {
  if (_conflictResolver === null) {
    _conflictResolver = new ConflictResolver(dbConnection, similarityThreshold);
  } else if (dbConnection != null && _conflictResolver.db === null) {
    _conflictResolver.setDbConnection(dbConnection);
  }
  return _conflictResolver;
}

/** Reset ConflictResolver singleton (for testing). */
export function resetConflictResolver(): void {
  _conflictResolver = null;
}
